runtime/drivers/duckdb/olap.go

```diff
@@ -181,6 +186,344 @@ func (c *connection) EstimateSize() (int64, bool) {
+	dbFile := filepath.Join(sourceDir, fmt.Sprintf("%s.db", newVersion))
+	db := dbName(name, newVersion)
+	// attach new db
+	err := c.Exec(ctx, &drivers.Statement{Query: fmt.Sprintf("ATTACH '%s' AS %s", dbFile, safeSQLName(db))})
+	if err != nil {
+		removeDBFile(dbFile)
+		return err
+	}
+
+	if err := c.Exec(ctx, &drivers.Statement{
+		Query:       fmt.Sprintf("CREATE OR REPLACE TABLE %s.default AS (%s)", safeSQLName(db), sql),
+		Priority:    1,
+		LongRunning: true,
+	}); err != nil {
+		_ = c.Exec(ctx, &drivers.Statement{Query: fmt.Sprintf("DETACH %s", safeSQLName(db)), Priority: 100})
+		removeDBFile(dbFile)
+		return err
+	}
+
+	return c.WithConnection(ctx, 1, true, false, func(ctx, ensuredCtx context.Context, _ *dbsql.Conn) error {
+		// create view query
+		err = c.Exec(ctx, &drivers.Statement{
+			Query: fmt.Sprintf("CREATE OR REPLACE VIEW %s AS SELECT * FROM %s.default", safeSQLName(name), safeSQLName(db)),
+		})
+		if err != nil {
+			_ = c.Exec(ensuredCtx, &drivers.Statement{Query: fmt.Sprintf("DETACH %s", safeSQLName(db)), Priority: 100})
+			removeDBFile(dbFile)
+			// not cleaning the source directory for graceful failure of repeate create or replace table statements
+			return err
+		}
+
+		if oldVersionExists {
+			oldDB := dbName(name, oldVersion)
+			// ignore these errors since source has been correctly ingested and attached
+			if err = c.Exec(ensuredCtx, &drivers.Statement{Query: fmt.Sprintf("DETACH %s", safeSQLName(oldDB))}); err != nil {
+				c.logger.Error("detach failed", zap.String("db", oldDB), zap.Error(err))
+			}
+			removeDBFile(filepath.Join(sourceDir, fmt.Sprintf("%s.db", oldVersion)))
+		}
+		return nil
+	})
+}
+
+// DropTable implements drivers.OLAPStore.
+func (c *connection) DropTable(ctx context.Context, name string, view bool) error {
+	if view {
+		return c.Exec(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("DROP VIEW IF EXISTS %s", safeSQLName(name)),
+			Priority:    100,
+			LongRunning: true,
+		})
+	}
+	if !c.config.TableAsView {
+		return c.Exec(ctx, &drivers.Statement{
+			Query:       fmt.Sprintf("DROP TABLE IF EXISTS %s", safeSQLName(name)),
+			Priority:    100,
+			LongRunning: true,
+		})
+	}
+
+	// ignore query cancellations
+	ctx = context.WithValue(context.Background(), connCtxKey{}, connFromContext(ctx))
+	sourceDir := filepath.Join(filepath.Dir(c.config.DBFilePath), name)
+	version, exist, err := c.tableVersion(name)
+	if err != nil {
+		return err
+	}
+
+	if !exist {
+		return nil
+	}
+	// drop view
+	err = c.Exec(ctx, &drivers.Statement{
+		Query:       fmt.Sprintf("DROP VIEW IF EXISTS %s", safeSQLName(name)),
+		Priority:    100,
+		LongRunning: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	oldDB := dbName(name, version)
+	err = c.Exec(ctx, &drivers.Statement{
+		Query:       fmt.Sprintf("DETACH %s", safeSQLName(oldDB)),
+		Priority:    100,
+		LongRunning: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	// delete source directory
+	return os.RemoveAll(sourceDir)
```

Comment:
1. Only necessary to ignore query cancellations after the first query
2. Is DETACH safe to call multiple times? If not, it should ignore the error to ensure idempotency

Reply:
2. No. I want the drop to fail if detach fails to release any resources with `attach`. I can add checks for db not found errors.

Reply:
Why is that? The reconcile logic assumes that table drops are idempotent, i.e. act like `DROP TABLE IF EXISTS`, not `DROP TABLE`. 

Imagine pod shutdown halfway through a table drop – when it restarts, reconcile will re-run incomplete operations.

Reply:
Yeah agreed. I am saying we should ignore `db not found error` to ensure `idempotency` but fail on other errors.